### PR TITLE
Take 2: Fix the race between configuring cbr0 and restarting static pods

### LINKED ISSUE #10122: kube-apiserver unable to write to etcd
In about 100 test runs overnight, 5 clusters failed with the same problem:
Nodes try to register with the apiserver, apiserver sends back a 500.
Looking at the kube-apiserver logs, I see:
```
E0619 20:37:50.685032       5 errors.go:62] apiserver received an error that is not an api.Status: 501: All the given peers are not reachable (failed to propose on members [http://127.0.0.1:4001] twice [last error: Unexp
ected HTTP status code]) [0]
I0619 20:37:50.685229       5 handlers.go:135] POST /api/v1/namespaces: (12.228060557s) 500
...
E0619 20:37:50.643558       5 errors.go:62] apiserver received an error that is not an api.Status: 501: All the given peers are not reachable (failed to propose on members [http://127.0.0.1:4001] twice [last error: Unexp
ected HTTP status code]) [0]
I0619 20:37:50.643762       5 handlers.go:135] POST /api/v1/nodes: (12.228247163s) 500
```

To me, that looks like writes to etcd are failing. Looking into the etcd container logs, I see a lot of:

```
2015/06/19 20:40:38 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:38 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:38 etcdserver: publish error: etcdserver: request timed out
2015/06/19 20:40:38 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:41 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:41 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:42 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:43 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:43 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:43 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:43 etcdserver: publish error: etcdserver: request timed out
2015/06/19 20:40:44 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:44 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:44 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:47 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:47 etcdhttp: unexpected error: etcdserver: request timed out
2015/06/19 20:40:47 etcdhttp: unexpected error: etcdserver: request timed out
```

Trying to hit etcd directly:
```
# curl -L http://127.0.0.1:4001/v2/keys/message -XPUT -d value="Hello world"
{"message":"Internal Server Error"}
```
Etcd's data directory is pretty empty, other than one entry in members/wal, so I tried listing members:
```
# curl http://127.0.0.1:4001/v2/members
{"members":[]}
```
In a working cluster, that looks more like:
```
# curl http://127.0.0.1:4001/v2/members
{"members":[{"id":"ce2a822cea30bfca","name":"default","peerURLs":["http://localhost:2380","http://localhost:7001"],"clientURLs":["http://127.0.0.1:4001"]}]}
```

I'm about at the extent of my etcd knowledge, so if anybody has any idea what else to poke at, I'd love to hear it.

### PR DESCRIPTION
**Pull request by dchen1107**
The pr rever-revert "Fix the race between configuring cbr0 and restarting static pods", also
- add iptables rule for MASQUERADE for egress
- make sure mtu=1460 for cbr0

With this pr, the new cluster I brought up, I checked both master and minion on network configurations:

On master node:

```
root@kubernetes-master:/var/log# iptables-save | grep MASQUERADE
-A POSTROUTING ! -d 10.0.0.0/8 -o eth0 -j MASQUERADE

# ip addr show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN 
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1460 qdisc pfifo_fast state UP qlen 1000
    link/ether 42:01:0a:f0:9c:cf brd ff:ff:ff:ff:ff:ff
    inet 10.240.156.207/32 brd 10.240.156.207 scope global eth0
       valid_lft forever preferred_lft forever
4: cbr0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1460 qdisc noqueue state UNKNOWN 
    link/ether f6:1f:bd:10:27:48 brd ff:ff:ff:ff:ff:ff
    inet 10.246.0.1/24 scope global cbr0
       valid_lft forever preferred_lft forever
```

On Minion node

```
root@kubernetes-minion-fptt:/var/log# iptables-save | grep MASQUERADE
-A POSTROUTING ! -d 10.0.0.0/8 -o eth0 -j MASQUERADE

root@kubernetes-minion-fptt:/var/log# ip addr show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN 
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1460 qdisc pfifo_fast state UP qlen 1000
    link/ether 42:01:0a:f0:4d:8c brd ff:ff:ff:ff:ff:ff
    inet 10.240.77.140/32 brd 10.240.77.140 scope global eth0
       valid_lft forever preferred_lft forever
4: cbr0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1460 qdisc noqueue state UP 
    link/ether 4a:83:23:02:49:e6 brd ff:ff:ff:ff:ff:ff
    inet 10.244.0.1/24 scope global cbr0
       valid_lft forever preferred_lft forever
6: vethe9346d5: <BROADCAST,UP,LOWER_UP> mtu 1460 qdisc noqueue master cbr0 state UP 
    link/ether 4a:83:23:02:49:e6 brd ff:ff:ff:ff:ff:ff
```

I did manual tests to verify internet connection from inside of container too;

```
# docker ps
CONTAINER ID        IMAGE                                      COMMAND                CREATED             STATUS              PORTS               NAMES
15157eba242a        redis:latest                               "/entrypoint.sh redi   33 seconds ago      Up 32 seconds                           k8s_master.718e46d0_redis-master-mhbst_default_7e7c1eae-1ac3-11e5-8010-42010af09ccf_163b0101                                        
0753208fa688        gcr.io/google_containers/pause:0.8.0       "/pause"               44 seconds ago      Up 43 seconds                           k8s_POD.49eee8c2_redis-master-mhbst_default_7e7c1eae-1ac3-11e5-8010-42010af09ccf_9ab956a2                                           

# docker exec -t -i 15157eba242a sh
# ping google.com
PING google.com (74.125.70.100): 48 data bytes
56 bytes from 74.125.70.100: icmp_seq=0 ttl=50 time=0.954 ms
56 bytes from 74.125.70.100: icmp_seq=1 ttl=50 time=0.809 ms
56 bytes from 74.125.70.100: icmp_seq=2 ttl=50 time=0.815 ms
...
```

Fix #10122 and #9822

### CODE REVIEW

**Comment by dchen1107** [comment]
ok to test

**Comment by dchen1107** [comment]
ok to test

**Comment by dchen1107** [comment]
cc/ @brendandburns  @thockin

**Comment by dchen1107** [comment]
ok to test

**Comment by brendandburns** [comment]
I only reviewed the commit that is the delta on top of the revert, but LGTM.

**Comment by dchen1107** [comment]
@roberthbailey 10.0.0.0/8 is hard coded in https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/saltbase/salt/_states/container_bridge.py#L51 

This time I did line-to-line comparison between container_bridge.py and cotnainer_bridge.go. That is why I found mtu difference too.

**Comment by dchen1107** [comment]
cc/ @quinton-hoole for another look too since @thockin is extremely busy here?

**Comment by quinton-hoole** [comment]
Sure, will do now.

**Comment by roberthbailey** [comment]
The difference is that before container_bridge.py was only run on salt-based systems. Since this is in the kubelet it will run on all systems.

**Comment by dchen1107** [comment]
Shouldn't this piece of code is still guarded by kubelet flag --configure-cbr0?

**Comment by dchen1107** [comment]
@quinton-hoole Yes. Kubelet periodically check network configurations.

**Comment by dchen1107** [comment]
@quinton-hoole  Renamed the method to more specific name you proposed. 

ok to test

File: cluster/gce/configure-vm.sh [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #!/bin/bash
+set -x
```
Comment (thockin) [inline cluster/gce/configure-vm.sh]: debug code?

**Comment by quinton-hoole** [comment]
LGTM for the delta, @thockin is also busy with a final review.

File: cluster/gce/configure-vm.sh (hunk 1 above)
Comment (dchen1107) [inline cluster/gce/configure-vm.sh]: I am adding this on purpose. configure-vm.sh runs on every gce instances at boot time. Once it went wrong, it is very difficult to debug. This only write to startupscript.log. I hope it is ok.

**Comment by dchen1107** [comment]
I add a TODO to use pkg/util/iptables too. 

ok to test

File: pkg/kubelet/container_bridge.go [hunk 2]
```diff
@@ -79,6 +113,21 @@ func cbr0CidrCorrect(wantCIDR *net.IPNet) bool {
...
 	}
 	cbr0CIDR.IP = cbr0IP
+
 	glog.V(5).Infof("Want cbr0 CIDR: %s, have cbr0 CIDR: %s", wantCIDR, cbr0CIDR)
 	return wantCIDR.IP.Equal(cbr0IP) && bytes.Equal(wantCIDR.Mask, cbr0CIDR.Mask)
 }
+
+func ensureIPTablesMasqRule() error {
+	// Check if the MASQUERADE rule exist or not
+	if err := exec.Command("iptables", "-t", "nat", "-C", "POSTROUTING", "-o", "eth0", "-j", "MASQUERADE", "!", "-d", "10.0.0.0/8").Run(); err == nil {
```
Comment (thockin) [inline pkg/kubelet/container_bridge.go]: TODO: use util.iptables

Comment (dchen1107) [inline pkg/kubelet/container_bridge.go]: Done already :-)

**Comment by thockin** [comment]
Need to update docs/networking.md (search for container_bridge) and docs/design/networking.md.

Should we delete cluster/saltbase/salt/docker/init.sls too?

File: cluster/gce/configure-vm.sh (hunk 1 above)
Comment (roberthbailey) [inline cluster/gce/configure-vm.sh]: /cc @zmerlynn 

any reason to this as a set rather than on the top line, e.g. `#!/bin/bash -x`

**Comment by dchen1107** [comment]
@thockin I filed issue #10321 for document updates. I am not sure about cluster/saltbase/salt/docker/init.sls since it set environment_file for different os family too. Also it still controls the docker flags

File: cluster/gce/configure-vm.sh (hunk 1 above)
Comment (zmerlynn) [inline cluster/gce/configure-vm.sh]: Oi .. as someone who looks at a ton of startupscript.logs, this is really spammy. :/ I worked for a while to make the boot script more concise (until the salt execution, but that's the last step), so 99% of the noise is now a bug.

Comment (dchen1107) [inline cluster/gce/configure-vm.sh]: @zmerlynn I heard you, and removed it. Thanks!

**Comment by dchen1107** [comment]
ok to test

**Comment by thockin** [comment]
I meant to say that we should delete the file ./cluster/saltbase/salt/_states/container_bridge.py but I cut-pasted the wrong filename.

LGTM

**Comment by mbforbes** [comment]
Jenkins passed... here we go...

**Pull request merged**